File: ios/Sources/ARYPush/Internal/DeviceInfoProvider.swift
import Foundation
import UIKit

/// Collects the small, documented set of device facts the backend needs to target a notification.
///
/// Deliberately narrow. Nothing here identifies a person, and nothing is read that would require
/// a permission or a privacy manifest entry beyond the SDK's declared use: no IDFA, no
/// `identifierForVendor`, no contacts, no location. Every field below is listed in
/// docs/SECURITY.md so that a privacy review can be done against the documentation rather than
/// against the source.
final class DeviceInfoProvider {

    /// The SDK version, kept in one place so it can never drift from the released tag.
    static let sdkVersion = "1.0.0"

    private let bundle: Bundle

    init(bundle: Bundle = .main) {
        self.bundle = bundle
    }

    /// Host application version, e.g. `5.2.0`.
    var appVersion: String? {
        bundle.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String
    }

    /// Host application build number.
    var appBuild: String? {
        bundle.object(forInfoDictionaryKey: "CFBundleVersion") as? String
    }

    var sdkVersion: String { Self.sdkVersion }

    /// iOS release version, e.g. `17.4`.
    var osVersion: String { UIDevice.current.systemVersion }

    /// Hardware identifier, e.g. `iPhone16,1`.
    ///
    /// `UIDevice.model` only ever returns "iPhone", which is useless for targeting, so the
    /// machine identifier is read from `uname` instead. It describes a model, not a device, and
    /// cannot be used to identify a person.
    var deviceModel: String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let identifier = withUnsafePointer(to: &systemInfo.machine) { pointer in
            pointer.withMemoryRebound(to: CChar.self, capacity: 1) { String(validatingUTF8: $0) }
        }
        return identifier?.trimmingCharacters(in: .whitespacesAndNewlines)
            ?? UIDevice.current.model
    }

    /// Current BCP-47 locale, e.g. `en-PK`. Read on each access so a change is picked up.
    var locale: String { Locale.current.identifier.replacingOccurrences(of: "_", with: "-") }

    /// IANA timezone, e.g. `Asia/Karachi`.
    var timezone: String { TimeZone.current.identifier }

    /// The host application's bundle identifier, which is also its public App Store identity.
    var bundleIdentifier: String { bundle.bundleIdentifier ?? "unknown" }
}

File: android/sdk/src/main/kotlin/com/ary/push/internal/device/DeviceInfoProvider.kt
package com.ary.push.internal.device

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.pm.PackageInfoCompat
import com.ary.push.BuildConfig
import java.util.Locale
import java.util.TimeZone

/**
 * Collects the small, documented set of device facts the backend needs to target a notification.
 *
 * Deliberately narrow. Nothing here identifies a person, and nothing is read that would require
 * a runtime permission: no advertising id, no hardware serial, no accounts, no phone number, no
 * location. Every field below is listed in docs/SECURITY.md so that a privacy review can be
 * done against the documentation rather than against the source.
 */
internal class DeviceInfoProvider(context: Context) {

    private val appContext = context.applicationContext

    /** Host application version name, e.g. `5.2.0`. */
    val appVersion: String? by lazy { packageInfo()?.versionName }

    /** Host application version code, as a string so that long codes survive JSON. */
    val appBuild: String? by lazy {
        packageInfo()?.let { PackageInfoCompat.getLongVersionCode(it).toString() }
    }

    /** ARY Push SDK version, taken from the build so it can never drift from the artifact. */
    val sdkVersion: String get() = BuildConfig.SDK_VERSION

    /** Android release version, e.g. `14`. */
    val osVersion: String get() = Build.VERSION.RELEASE ?: Build.VERSION.SDK_INT.toString()

    /** Marketing device model, e.g. `Pixel 8`. */
    val deviceModel: String get() = "${Build.MANUFACTURER} ${Build.MODEL}".trim()

    /** Current BCP-47 locale, e.g. `en-PK`. Read on each access so a locale change is picked up. */
    val locale: String get() = Locale.getDefault().toLanguageTag()

    /** IANA timezone, e.g. `Asia/Karachi`. */
    val timezone: String get() = TimeZone.getDefault().id

    /** The host application's package name, which is also its public Play identity. */
    val packageName: String get() = appContext.packageName

    private fun packageInfo() = try {
        appContext.packageManager.getPackageInfo(appContext.packageName, 0)
    } catch (e: PackageManager.NameNotFoundException) {
        null
    }
}
